libs/entitlement/tools/securitymanager-scanner/src/main/java/org/elasticsearch/entitlement/tools/securitymanager/scanner/SecurityCheckClassVisitor.java

```diff
@@ -0,0 +1,262 @@
+        private String runtimePermissionType;
+        private final int methodAccess;
+
+        protected SecurityCheckMethodVisitor(MethodVisitor mv, String methodName, int methodAccess) {
+            super(ASM9, mv);
+            this.methodName = methodName;
+            this.methodAccess = methodAccess;
+        }
+
+        private static final Set<String> KNOWN_PERMISSIONS = Set.of("jdk.vm.ci.services.JVMCIPermission");
+
+        @SuppressForbidden(reason = "System.err is OK for this simple command-line tool")
+        private void handleException(String className, Throwable e) {
+            System.err.println("Cannot process " + className + ": " + e.getMessage());
+        }
+
+        @Override
+        public void visitTypeInsn(int opcode, String type) {
+            super.visitTypeInsn(opcode, type);
+            if (opcode == NEW) {
+                if (type.endsWith("Permission")) {
+                    var objectType = Type.getObjectType(type);
+                    if (KNOWN_PERMISSIONS.contains(objectType.getClassName())) {
+                        permissionType = type;
+                    } else {
+                        try {
+                            var clazz = Class.forName(objectType.getClassName());
+                            if (Permission.class.isAssignableFrom(clazz)) {
+                                permissionType = type;
+                            }
+                        } catch (ClassNotFoundException e) {
+                            handleException(objectType.getClassName(), e);
+                        }
+                    }
+                }
+            }
+        }
+
+        @Override
+        @SuppressForbidden(reason = "We need to violate java's access system to access private parts")
+        public void visitFieldInsn(int opcode, String owner, String name, String descriptor) {
+            super.visitFieldInsn(opcode, owner, name, descriptor);
+            if (opcode == GETSTATIC && descriptor.endsWith("Permission;")) {
+                var permissionType = Type.getType(descriptor);
+                if (permissionType.getSort() == Type.ARRAY) {
+                    permissionType = permissionType.getElementType();
+                }
+                try {
+                    var clazz = Class.forName(permissionType.getClassName());
+                    if (Permission.class.isAssignableFrom(clazz)) {
+                        this.permissionType = permissionType.getInternalName();
+                    }
+                } catch (ClassNotFoundException e) {
+                    handleException(permissionType.getClassName(), e);
+                }
+
+                var objectType = Type.getObjectType(owner);
+                try {
+                    var clazz = Class.forName(objectType.getClassName());
+                    Arrays.stream(clazz.getDeclaredFields())
+                        .filter(f -> Modifier.isStatic(f.getModifiers()) && Modifier.isFinal(f.getModifiers()))
```

Comment:
What is this filter for? Why do we need to avoid non-static and non-final fields?

Reply:
This is also a "optional additional info" bit: it covers a pattern widely used in the JDK where there is a constant field (static final) declaring a permission.
It is used to try and extract the permission type and name (see readme). If not, fair enough, we will _not_ skip this, just have less information to use when reviewing the list.